Assets/MRTK/SDK/Experimental/Dock/Dockable.cs

```diff
@@ -0,0 +1,335 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.MixedReality.Toolkit.UI;
+using Microsoft.MixedReality.Toolkit.Utilities.Solvers;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Assertions;
+
+namespace Microsoft.MixedReality.Toolkit.Experimental.UI
+{
+    /// <summary>
+    /// The possible states of a <see cref="Dockable"/> object.
+    /// </summary>
+    public enum DockingState
+    {
+        Undocked,
+        Docking,
+        Docked,
+        Undocking
+    }
+
+    /// <summary>
+    /// Add a Dockable component to any object that has a <see cref="Dockable"/> and an <see cref="ObjectManipulator"/>
+    /// or <see cref="ManipulationHandler"/> to make it dockable in Docks. That allows this object to be used
+    /// as part of a palette, shelf or navigation bar together with other objects.
+    /// </summary>
+    /// <seealso cref="Dock"/>
+    /// <seealso cref="DockPosition"/>
+    [AddComponentMenu("Scripts/MRTK/Experimental/Dock/Dockable")]
+    public class Dockable : MonoBehaviour
+    {
+        [Experimental]
+        [SerializeField, ReadOnly]
+        [Tooltip("Current state of this dockable in regards to a dock.")]
+        private DockingState dockingState = DockingState.Undocked;
+
+        [SerializeField]
+        [Tooltip("Time to animate any move/scale into or out of the dock.")]
+        private float moveLerpTime = 0.1f;
+
+        [SerializeField]
+        [Tooltip("Time to animate an element when it's following the dock (use 0 for tight attachment)")]
+        private float moveLerpTimeWhenDocked = 0.05f;
+
+        /// <summary>
+        /// True if this object can currently be docked, false otherwise.
+        /// </summary>
+        public bool CanDock => dockingState == DockingState.Undocked || dockingState == DockingState.Undocking;
+
+        /// <summary>
+        /// True if this object can currently be undocked, false otherwise.
+        /// </summary>
+        public bool CanUndock => dockingState == DockingState.Docked;
+
+        // Constants
+        private static readonly float distanceTolerance = 0.01f; // in meters
+        private static readonly float angleTolerance = 3.0f; // in degrees
+        private static readonly float scaleTolerance = 0.01f; // in percentage
+
+        private DockPosition dockedPosition = null;
+        private Vector3 dockedPositionScale = Vector3.one;
+
+        private HashSet<DockPosition> overlappingPositions = new HashSet<DockPosition>();
+        private Vector3 originalScale = Vector3.one;
+        private bool isDragging = false;
+        private ObjectManipulator objectManipulator;
+        private ManipulationHandler manipulationHandler;
+
+        /// <summary>
+        /// Subscribes to manipulation events.
+        /// </summary>
+        private void Start()
+        {
```

Comment:
I'd put this in `OnEnable` to ensure every time the object is activated it will work as expected.